examples/hetero/bipartite_sage_unsup.py

```diff
@@ -0,0 +1,278 @@
+    ).relu()
+    user_x = self.conv3(
+      (item_x, user_x),
+      edge_index_dict[('item', 'rev_to', 'user')],
+    ).relu()
+    return self.lin(user_x)
+
+
+class EdgeDecoder(torch.nn.Module):
+  def __init__(self, hidden_channels):
+    super().__init__()
+    self.lin1 = Linear(2 * hidden_channels, hidden_channels)
+    self.lin2 = Linear(hidden_channels, 1)
+
+  def forward(self, z_src, z_dst, edge_label_index):
+    row, col = edge_label_index
+    z = torch.cat([z_src[row], z_dst[col]], dim=-1)
+
+    z = self.lin1(z).relu()
+    z = self.lin2(z)
+    return z.view(-1)
+
+
+class Model(torch.nn.Module):
+  def __init__(self, num_users, num_items, hidden_channels, out_channels):
+    super().__init__()
+    self.user_emb = Embedding(num_users, hidden_channels, device=device)
+    self.item_emb = Embedding(num_items, hidden_channels, device=device)
+    self.item_encoder = ItemGNNEncoder(hidden_channels, out_channels)
+    self.user_encoder = UserGNNEncoder(hidden_channels, out_channels)
+    self.decoder = EdgeDecoder(out_channels)
+
+  def forward(self, x_dict, edge_index_dict, edge_label_index):
+    z_dict = {}
+    x_dict['user'] = self.user_emb(x_dict['user'])
+    x_dict['item'] = self.item_emb(x_dict['item'])
+    z_dict['item'] = self.item_encoder(
+        x_dict['item'],
+        edge_index_dict[('item', 'to', 'item')],
+    )
+    z_dict['user'] = self.user_encoder(x_dict, edge_index_dict)
+
+    return self.decoder(z_dict['item'], z_dict['user'], edge_label_index)
+
+
+model = Model(
+  num_users=data['user'].num_nodes,
+  num_items=data['item'].num_nodes,
+  hidden_channels=64,
+  out_channels=64,
+).to(device)
+optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
+
+
+def train():
+  model.train()
+
+  total_loss = total_examples = 0
+  for batch in tqdm.tqdm(train_loader):
+    batch = batch.to(device)
+    optimizer.zero_grad()
+
+    pred = model(
+      batch.x_dict,
+      batch.edge_index_dict,
+      batch['item', 'rev_to', 'user'].edge_label_index,
+    )
+    loss = F.binary_cross_entropy_with_logits(
+      pred, batch['item', 'rev_to', 'user'].edge_label)
+
+    loss.backward()
+    optimizer.step()
+    total_loss += float(loss)
+    total_examples += pred.numel()
+
+  return total_loss / total_examples
+
+
+@torch.no_grad()
+def test(loader):
+  model.eval()
+
+  preds, targets = [], []
+  for batch in tqdm.tqdm(loader):
+    batch = batch.to(device)
+
+    pred = model(
+      batch.x_dict,
+      batch.edge_index_dict,
+      batch['item', 'rev_to', 'user'].edge_label_index,
+    ).sigmoid().view(-1).cpu()
+    target = batch['item', 'rev_to', 'user'].edge_label.long().cpu()
+
+    preds.append(pred)
+    targets.append(target)
+
+  pred = torch.cat(preds, dim=0).numpy()
+  target = torch.cat(targets, dim=0).numpy()
+
+  return roc_auc_score(target, pred)
+
+
+for epoch in range(1, 21):
+  loss = train()
+  test_auc = test(test_loader)
+
+  print(f'Epoch: {epoch:02d}, Loss: {loss:4f}, '
```

Comment:
How do the model AUC and training performance compare to PyG's？

Reply:
Performance of GLT on one V100 16G GPU 16G with the same default parameter setting is about 3.1x speedup compared to PyG's and AUC is close to PyG's.